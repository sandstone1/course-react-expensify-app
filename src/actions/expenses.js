
//  we will need the uuid import for the AddExpense action generator
import uuid from 'uuid';
// see -- Mark 1 -- below for details
import database from '../firebase/firebase';
// we will need to set up the named exports
// format is " export const action generator name "



// ==============================
// ACTION GENERATOR #1
// ==============================


// we will destructure the first argument and if it doesn't exist we will destructure
// an empty object
// we can define all the things we want within the first {} within the argument
// we want the user to provide the value for description and note but if the user
// does not provide a value then an empty string will surfice
// next up, we default the amount to an appropriate value like 0
// next up, we default the createdAt to an appropriate value like 0 until we talk about
// timestamps
export const addExpense = ( expense ) => ( {
    type : 'ADD_EXPENSE',
    expense : expense
} );



// ==============================
// ACTION GENERATOR #2
// ==============================


// we need the id from the expense we are going to remove
export const removeExpense = ( { id } = {} ) => ( {
    type : 'REMOVE_EXPENSE',
    expense : {
        id : id
    }
} );




// ==============================
// ACTION GENERATOR #3
// ==============================


// pass in our 2 arguments: id and updates and neither need default values since we
// we don't have the ID and were not updating anything ( don't understand this last
// part ) 
export const editExpense = ( id, updates ) => ( {
    type : 'EDIT_EXPENSE',
    id, // remember this is equal to id : id
    updates // remember this is equal to updates : updates
} );


// -- Mark 1 --
// lecture 152: Asynchronous Redux Actions
// addExpense will dispatch an object and that object will change the redux store whereas
// startAddExpense is just going to start that process off and startAddExpense will dispatch
// addExpense inside the function below and that dispatch will change the redux store and in
// the past we always returned objects but in this case we are going to return a function and
// this function works because we set up the middleware for redux-thunk and by default doing
// something like this would not work

// the return function below gets called internally by redux and it gets called with dispatch
// as the argument and this gives us access to dispatch inside our return function and we
// use dispatch after we write some data to Firebase and after waiting for that data to correctly
// sync and then after that we will use dispatch to dispatch addExpense and we will make sure the
// redux store reflects those changes as well and if we going to be saving to Firebase inside the
// return function we will need to restructure addExpense above so instead of having the
// default in the addExpense action generator we want to put the default key value pairs inside
// the return function and will switch up how we set those defaults in the argument to
// startAddExpense

// we will call the argument to startAddExpense " expenseData " and by default set it equal to an
// empty object and set up a const inside the return function and destructure a bunch of stuff and
// we'll destructure from expenseData and then copy the addExpense argument and place it inside the
// const object below and this is exactly the same argument as we had in the addExpense action
// generator and since we are setting up the defaults below that means addExpense can get a whole
// lot simplier

// the next thing we are going to do is use push to save some data to Firebase and we first need
// to import the database variable into this file so we will do this above:
// " import database from '../firebase/firebase'; " and then we can use the database variable below
// and we will get a reference to " expenses " and then we will push an expense, which is an object,
// into our Firebase database and we will create an variable for this object
// and call it expense and then pass that variable into push()
// and then we can attach the then call and at this point what is going to happen? the function
// will run and then we will take all the data or expenseData and save the data to Firebase and at
// this point there is still an important step left and that step is to dispatch the addExpense
// action generator from above and change the addExpense action generator above to this:
/*
export const addExpense = ( expense ) => ( {
    type : 'ADD_EXPENSE',
    expense : expense
} );
*/

// and yes we broke some test cases and we will fix those in the next video but for now let's focus
// on getting this functionality working and what are we trying to add as the argument to the
// addExpense call below? we are trying to add the expense we just created in the expense variable
// up above and we have to include the ID as well in the addExpense call and the good news for us
// is that the then callback or success case gets called with a reference and we can access that
// reference by including the reference as the argument to then() and we'll call the reference
// " ref " and we can access the ref and do something meaningful with it and in our case we'll use
// ref to get the id by creating an object inside the addExpense call and defining the ID as
// follows: " id : ref.key " and then we'll tack on all the properties in expense by using
// the spread operator or " ...expense "
export const startAddExpense = ( expenseData = {} ) => {

    return ( dispatch ) => {

        const {
            description = '',
            note        = '',
            amount      = 0,
            createdAt   = 0
        } = expenseData;

        const expense = { description, note, amount, createdAt };

        // add return below and see -- Mark 1 -- of lecture 153: Testing Async Redux Actions:
        // Part 1 for details on why we added the return keyword below
        return database.ref( 'expenses' ).push( expense ).then( ( ref ) => {
            
            dispatch( addExpense( { 
                id : ref.key,
                ...expense
            } ) );
        
        } );

    };

};

// and there we have it, this is all we need to do inside expenses.js and the last step before this
// is going to work is to make sure we dispatch startAddExpense as opposed to dispatching
// addExpense and

// -- go to the AddExpensePage component page --

// and change:
// " import { addExpense } from '../actions/expenses'; " to
// " import { startAddExpense } from '../actions/expenses'; ""

// and change:
// " addExpense : ( expense ) => dispatch( addExpense( expense ) ) " to
// " startAddExpense : ( expense ) => dispatch( startAddExpense( expense ) ) "

// and change:
// " this.props.addExpense( expense ); " to 
// " this.props.startAddExpense( expense ); "

// with these changes in place we are all done and we can test this out and test this out by
// creating an expense and what we hope to see is the expense showing up in the browser meaning it
// was added to redux and we are also hoping the expense shows up in our Firebase database and
// after I add an expense in the expensify site I get redirected to the home page and I see my
// expense listed meaning it was added to the redux store and when I go to Firebase I see our
// new expense listed in the database

// so by using redux-thunk were going to be able to create asynchronous actions and these actions
// will do something first soemthing asynchronous like make a Firebase data call and then these
// actions will use dispatch to actually change the redux store

// remember, this is just one of our 4 CRUD operations so we have create done but were still not
// reading from Firebase but if we do a page refresh we lose the data and this is expected for the
// moment and we'll read from Firebase in just a little bit but what I want to focus on in the
// next video though is how we can test the code we just wrote in this video and we want to make
// sure that everything we expect to happen actually happens and we will change some test cases
// for things we changed in this video and remember we changed the addExpense action generator so
// we have some breaking test cases there and we changed the AddExpensePage component so we will
// need to make some changes to that test case as well


// -- END OF MARK 1 --
