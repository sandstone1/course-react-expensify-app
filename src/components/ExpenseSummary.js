
// we need React since we are using JSX
import React               from 'react';
import { connect }         from 'react-redux';
import numeral             from 'numeral';
import selectExpenses      from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expenses-total';



// create a stateless functional component and no need to manage any state in this component
// for the argument, we used desstructuring but could have done this:
// " ( props.expensesCount, props.expensesTotal ) "
export const ExpenseSummary = ( { expenseCount, expenseTotal } ) => {

    // above retrun() the 2 props: expenseCount and expenseTotal do not need to be enclosed within
    // {} but within return() the 2 props and all variables need to be enclosed in curly braces
    // or {}
    const expenseWord = expenseCount === 1 ? 'expense' : 'expenses';

    const totalFormatted = numeral( expenseTotal / 100 ).format( '$0,0.00' );

    // return some jsx that will be rendered to the screen
    return (
        <div>
            <h1>Viewing { expenseCount } { expenseWord } totaling { totalFormatted }</h1>
        </div>
    );
};



// we are going to get access to the entire state via the first argument
const mapStateToProps = ( state ) => {

    const visibleExpenses = selectExpenses( state.expenses, state.filters );

    // we will return our props: expenseCount and expenseTotal
    return {
        expenseCount : visibleExpenses.length,
        expenseTotal : selectExpensesTotal( visibleExpenses )
    };
  
};


// this gets exported to the ExpenseDashboardPage.js component, which includes
// the <ExpenseListFilters /> component within the ExpenseDashboardPage.js component
export default connect( mapStateToProps )( ExpenseSummary )
