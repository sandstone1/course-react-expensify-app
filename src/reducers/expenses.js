

// there are no dependencies below and we are not using any sort of library
// so there is no need to import anything and we are not using jsx below so
// there is no need to import react



// EXPENSES REDUCER
// pass in the 2 arguments: state and action and return the new state
// the default state value will be an empty array and we can create a
// variable for our default state value
const expensesReducerDefaultState = [];

// USE A DEFAULT EXPORT BELOW
// we will use a default export since there is one thing we want to export from
// this file
// for the default export, as the first line below, we could do:
// " const expensesReducer = ( state = expensesReducerDefaultState, action ) => { "
// and then " export default expensesReducer; "
// or we do
// " export default ( state = expensesReducerDefaultState, action ) => { "

export default ( state = expensesReducerDefaultState, action ) => {
    // we will switch what we do based off a particular value
    switch ( action.type ) {
        // inside the {} we can define the various cases we want to handle
        case 'ADD_EXPENSE':
            // here we want to run some code that adds the expense onto the array
            // don't want to use state.push( action.expense ) since push will change
            // the original array and we want to avoid this so we will use
            // state.concat( action.expense ) since this takes the state array and combines
            // it with action.expense to create a new array and we return this array
            /*
            return state.concat( action.expense );
            */
            
            // SPREAD OPERATOR
            // now the above works but we want to explore the ES6 spread operator and
            // this allows us to get the same thing done and more
            // as an example, const names = [ 'Andrew', 'Steve' ];
            // if I use the spread operator or [ ...names  ], the result is [ 'Andrew', 'Steve' ]
            // if I want to add an item onto the names array and return a new array, I do as
            // follows [ ...names, 'Mike' ] and the result is [ 'Andrew', 'Steve', 'Mike' ]
            // the great thing here is the spread operator did not change the names array
            // could do [ 'Adam', ...names, 'Mike' ] and the result is [ 'Adam', Andrew', 'Steve', 'Mike' ]
            // so instead of doing this " return state.concat( action.expense ); ", were going to do
            // return [ ...state, action.expense ]; ( so were taking the current state array and adding
            // on the action.expense object without changing the state array )
            return [ ...state, action.expense ];

        case 'REMOVE_EXPENSE':
            // use filter to remove an expense from the state array
            // filter creates a brand new array from another array and does this by looking at each item in the array
            // and will call a function once for each item in the array and the function will return true or false
            // if the function returns true then the current item will be passed into the new array
            // if the function returns false then the current item will not be passed into the new array

            // remember, action.expense.id is equal to expenseOne.expense.id
            // could use destructuring here so { id } = item (remember we have an object here) so we can pull
            // id off each item in the state array
            // BELOW WORKS
            // return state.filter( ( { id } ) => id !== action.expense.id )
            // BELOW WORKS
            return state.filter( ( item ) => item.id !== action.expense.id )

        case 'EDIT_EXPENSE':
            // will use the object spread operator below
            // what do we want to do here? we want to go through every singe expense in the array and
            // find the match and when we find the match then go ahead and correctly change the match
            // we going to use a some conditional logic to change the expense when our item.id matches
            // the action.id
            return state.map( ( item ) => {
                if ( item.id === action.id ) {
                    // if we have a match, we want to return a brand new object
                    return {
                        // will use the object spread operator below
                        // we will spread out the expense item so we want to grab description,
                        // note, amount and createdAt and we just want to override the previous values
                        // with the ones that the user passed in
                        // ...item will give us current item / expense key value pairs and ..action.updates
                        // will provide new key value pairs which will override the previous key value
                        // pairs and this will result in new key value pairs for this item / expense
                        // this results in a new key value pair or " amount : 500 "
                        ...item,
                        ...action.updates

                        // by using the object spread operator we have simple way to create a new object
                        // from an existing oject and we will be using the object spread operator
                        // extensively throughout the course
                    }
                }
                else {
                    // do nothing
                    return item;
                }
            } );

        // -- Mark 1 --
        // lecture 157: Fetching Expenses: Part 1
        case 'SET_EXPENSES':

            // all we need to do is return action.expenses and we are accounting for whatever
            // is in the state and we do not care about the previous expenses and this case
            // is designed to set the expenses array completely and now let's go to the
            // tests/reducers/expenses.test.js file and set the test case
            return action.expenses;

        // End of -- Mark 1 --


        // set up the default case
        default :
            // if the store is running for the first time, then the default state value
            // will be shown
            return state;
    }
};
