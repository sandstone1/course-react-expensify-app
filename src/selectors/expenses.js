
// -- Mark 1 --
// lecture 110: Filtering by Dates
import moment from 'moment';
// -----------

// there are no dependencies below and we are not using any sort of library
// so there is no need to import anything and we are not using jsx below so
// there is no need to import react


// get visible expenses
// will take 2 arguments: the expenses array and the filters object and using
// these 2 pieces of data, we can calculate what the visible expenses should
// look like

// startDate and endDate are timestamps and are in milliseconds
// timestamps start at January 1, 1970 ( known as the unix epoch )
// we will be using timestamps in this course as a way to store
// timezone independent time data

// USE A DEFAULT EXPORT BELOW
// we will use a default export since there is one thing we want to export from
// this file
// for the default export, as the first line below, we could do:
// " const getVisibleExpenses = ( expenses, { text, sortBy, startDate, endDate } ) => { "
// and then " export default getVisibleExpenses; "
// or we could do
// " export default ( expenses, { text, sortBy, startDate, endDate } ) => { "

// change filters to text, sortBy, startDate, endDate
export default ( expenses, { text, sortBy, startDate, endDate } ) => {
    // use filter to
    // filter creates a brand new array from another array and does this by looking at
    // each item in the array and will call a function once for each item in the array
    // and the function will return true or false
    // if the function returns true then the current item will be passed into the new array
    // if the function returns false then the current item will not be passed into the new array
    return expenses.filter( ( expense ) => {
        // we will make 3 const to store whether or not we have a match
        // for text, if I enter in some search text I want to show the expenses that actually
        // use that text value somewhere
        // for our purposes, we are just going to check the description and not the notes

        // challange for textMatch
        // 1 -- figure out if expense.description is inside the text variable string
        // 2 -- use the includes method and convert both strings to lower case
        const textMatch      = expense.description.toLowerCase().includes( text.toLowerCase() );
    
        // -- Mark 1 --
        // lecture 110: Filtering by Dates
        // first, we need to figure how to make the comparisons below using moment
        // and to do this we are going to be using the Query methods from moment
        // comment out the following startDateMatch and endDateMatch
        // const startDateMatch = typeof startDate !== 'number' || expense.createdAt >= startDate;
        // const endDateMatch   = typeof endDate   !== 'number' || expense.createdAt <= endDate;
        // startDateMatch will always be true if there is no startDate because we will never filter
        // out an expense when there is no startDate or the same goes for endDate
        // create a new moment to use as the argument to isSameOrBefore()
        // we passed in " day " as the second argument and therefore we will use day as the unit
        // of measurement so is startDate the same day or a day or days before the createdAtMoment
        const createdAtMoment = moment( expense.createdAt );
        const startDateMatch = startDate ? startDate.isSameOrBefore( createdAtMoment, 'day' ) : true ;
        const endDateMatch   = endDate ? endDate.isSameOrAfter( createdAtMoment, 'day' ) : true;
        // if all three const matches from above are true then we have a complete match and we want
        // to return true from filter and in this case the expense will be keeped ( or be passed into
        // the new array of expenses )
        // if we do not have a complete match then the expense will go away

        // so by setting up the DateRangePicker were going to be able to filter our expenses via this
        // calender widget and that's going to make searching and navigating our expenses much easier
        // we have 2 ways to filter: by date or description and we have 1 way to sort: by date or amount
        // recap: we changed startDate and endDate default state values to moments in time in the filter
        // reducer and we added the RangeDatePicker in the ExpenseListFilters component and the last thing
        // we changed we how we compared the values in startDateMatch and endDateMatch from above
        // ------------
        return textMatch && startDateMatch && endDateMatch;

    // now it's time to add sorting so we can get visible expenses
    // we want to be able to sort by date or sort by amount

    // we are going to look at 2 expenses and determine which one should come
    // before the other

    // sort gets called on an array and returns an array
    // when need to write a compare function or arguments ( a, b ) and then
    // compare them to determine which one comes first
    } ).sort( ( a, b ) => {
        // first thing we need to figure out are we sorting by date or amount
        if ( sortBy === 'date'  ) {
            // if we return -1 a would come first or if we return 1 b would come first
            // if a.createdAt < b.createdAt then we will show b first or have the
            // most recent expenses at the top of the list
            // use the ternary operator and if a.createdAt < b.createdAt then the
            // truthy value will be 1 ( b comes first ) otherwise a comes first
            return a.createdAt < b.createdAt ? 1 : -1;
        }
        // challange
        // sort by amount and put the one with a greater amount first
        else if ( sortBy === 'amount' ) {
            return a.amount < b.amount ? 1 : -1;
        }
    });
};
