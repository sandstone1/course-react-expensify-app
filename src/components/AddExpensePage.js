
// IMPORTS
// we need React since we are using JSX
import React from 'react';
// -- Mark 3 --
import ExpenseForm from './ExpenseForm';
// -- Mark 5 --
import { connect } from 'react-redux';
// -- Mark 5 --
import { startAddExpense } from '../actions/expenses';


// -- MARK 2 --
// example component 2
// this is stateless functional component
// way #1 - shorthand version
// took out {}; and the return statement so now implicitly returning jsx

// -- Mark 3 --
// add the ExpenseForm component below

// -- Mark 4 --
// here we pass the data up from the ExpenseForm component and this will
// allow us to determine what to do with the data when the user submits the
// form and do so on a dynamic basis so for this component we can dispatch
// the addExpense action and for the EditExpensePage component we can dispatch
// the editExpense action generator

// below we will provide a single prop or onSubmit and the onSubmit function
// gets called when we submit the form with valid data and in this case we
// will get the expense object back with all the properties we would expect
// ( i.e. description, note, amount and createdAt ) and we created these properties
// and their values in the ExpenseForm component and so we created the expense
// object below in the ExpenseForm component and did so with this.props.onSubmit({});


// ==============================
// GO TO EXPENSEFORM.JS -- Mark 5 --
// ==============================


// -- Mark 5 --
// the AddExpensePage component needs to be able to dispatch the given action
// to the redux store and this means we need to import the named export connect
// and remember we want to connect the AddExpensePage component to the store
// so we can dispatch
// we also have to import in the addExpense action generator in order to be able
// to use it below

// the last thing were going to do in this video is automate the redirect to the
// Dashboard page or the ExpenseDashboardPage component so when we submit the add
// expense form we get redirected to the Dashbaord page automatically
// how do we do the redirect? answer => the components we render inside react router
// get access to a bunch of special props; for example, if I use react dev tool tab
// I can see that the AddExpensePage component gets access to the props: dispatch,
// history, location and match
// and what we are going to do is use one of the history methods called push and
// push is how we can programatically change pages and push does take a single
// argument which is the path so if we wanted to go to the help page we would type
// ' /help ' so below we type " props.history.push( '/' ); " to be redirected
// to the dashboard page and our redirect will use browser routing and will not
// result in a full page refresh so now test that the component below works
// and with the push method a user can use the back button to go back a page
// so the push method is just like if a user clicked on a link

// recap: we validation, we have our date picker, we have automatic redirect
// and we have a reusable component in the ExpenseForm component
// also, we passed the data ( description, note, amount and createdAt ) out of
// the ExpenseForm component and into the AddExpensePage component and we did
// that by calling a prop that gets passed in from the parent ( i.e. onSubmit )
// and we called the prop inside the ExpenseForm component by typing
// " this.props.onSubmit({  }); " and we used " this.props.onSubmit " because
// the ExpenseForm component is a class based component and not a stateless
// functional component and we did it this way because we want to reuse the
// ExpenseForm component on both the AddExpensePage and the EditExpensePage and
// since these pages dispatch different actions we've extracted that away from
// the ExpenseForm component and put the dispatch call into the AddExpensePage
// and the EditExpensePage

// -------------

// -- Mark 7 --
// new class based component
// see below for details
// we switched to a class based component since we no longer want to define the onSubmit
// function inline
export class AddExpensePage extends React.Component {
    // set onSubmit to equal an arrow function and change
    // props to this.props
    onSubmit = ( expense ) => {
        // -- Mark 6 --
        // see below for details
        // props.dispatch( addExpense( expense ) );

        // -- Mark 8 --
        // lecture 125: Testing EditExpensePage
        // need to change the " onSubmit " property below to " addExpense "
        // GO TO -- Mark 3 -- in AddExpensePage.test.js
        this.props.startAddExpense( expense );
        this.props.history.push( '/' );
        // END OF -- Mark 8 --
    }
    // now all we need to do is define render and we return everything in the old component
    // starting at <div></div> and change onSubmit to " onSubmit={ this.onSubmit } " and the last
    // thing we need to do is test the unconnected version of the AddExpensePage component
    // so we will change " class AddExpensePage extends React.Component { " to
    // " export class AddExpensePage extends React.Component { " and now we have the unconnected
    // version above that we can test and we have the connected version below for our real app
    render() {
        return (
            <div>
                <h1>Add Expense</h1>
                <ExpenseForm
                    onSubmit={ this.onSubmit }
                />
            </div>
        );
    }
}

// -- END OF MARK 7 --


// old stateless functional component
/*
const AddExpensePage = ( props ) => (
    <div>
        <h1>Add Expense</h1>
        <ExpenseForm
            onSubmit={ ( expense ) => {
                // -- Mark 6 --
                // see below for details
                // props.dispatch( addExpense( expense ) );
                props.onSubmit( expense );
                props.history.push( '/' );
            } }
        />
    </div>
);
*/

// way #2 - longhand version
/*
const ExpenseDashBoardPage = () => {
    return (
        <div>
            some text
        </div>
    );
};
*/

// -- Mark 5 --
// we don't need anything from the state so we can leave the first () empty
// and now that we connected the AddExpensePage component to the store we have
// access to props.dispatch so above we need to add the props object as the
// argument to the AddExpensePage component above
// export default connect()( AddExpensePage );


// -- Mark 6 --
// lecture 124: Testing AddExpense Page
// comment out the Mark 5 export default and use the one below
// set up mapDispatchToProps and mapDispatchToProps is pretty similar to mapStateToProps
// but instead of working with the state it is working with dispatch and mapDispatchToProps
// gets called with dispatch as the argument so we access to use dispatch inside the function
// and the goal is return an object and on here we define various props and these props are
// going to call dispatch so at the end of the day we will be able to extract away
// " props.dispatch( addExpense( expense ) ); " from the AddExpensePage component and will
// replace it with " props.onSubmit( expense ) " and this will be much easier to test and to
// get " props.onSubmit( expense ) " to work we have to set up onSubmit down below and we will
// take expense in as an argument and then we will pass it through to 
// " dispatch( addExpense( expense ) ) " so now we have the exact same functionality but now
// we have a component that is more testable than ever and now we need to pass mapDispatchToProps
// in as the second argument to connect() below and mapDispatchToProps is a way to return your
// dispatch functions and abstract them away from the component itself and now let's convert
// our stateless function component above to a class based component and go to
// -- Mark 7 -- above

// -- Mark 8 --
// lecture 125: Testing EditExpensePage
// need to change the " onSubmit " property below to " addExpense " or we need to set the
// property equal to the name of the action generator or addExpense
const mapDispatchToProps = ( dispatch ) => ( {
    startAddExpense : ( expense ) => dispatch( startAddExpense( expense ) )
} );
// END OF -- Mark 8 --

// way #2 - longhand version
/*
const mapDispatchToProps = ( dispatch ) => {
    return {
        onSubmit : ( expense ) => dispatch( addExpense( expense ) )
    };
};
*/

export default connect( undefined, mapDispatchToProps )( AddExpensePage );

// -- END OF MARK 6 --
