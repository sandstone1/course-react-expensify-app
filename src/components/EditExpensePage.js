
// IMPORTS
// we need React since we are using JSX
import React from 'react';

// -- Mark 4 --
// step 1 is to connect the EditExpensePage component to the
// redux store so import the named export connect
import { connect } from 'react-redux';
// -----------

// -- Mark 5 --
// now that the expense is accessible, we can take things one step further
// and render the ExpenseForm component within the EditExpensePage component
import ExpenseForm from './ExpenseForm';
import { editExpense } from '../actions/expenses';
// ------------

// -- Mark 7 -- from ExpenseListItem
import { removeExpense } from '../actions/expenses';
// ------------

// -- MARK 3 --
// example component 3
// this is stateless functional component
// way #1 - shorthand version
// took out {}; and the return statement so now implicitly returning jsx
// -----------


// -- Mark 8 --
// lecture 125: Testing EditExpensePage
// first, refactor the ExitExpensePage component to be a class based component
// and this will allow us to change the inline functions to methods inside the class based
// component and therefore, we will not need to redefine every single time the component
// gets rendered
// second, we are going to set up mapDispatchToProps and in this case, we will be setting
// up 2 things: exitExpense and removeExpense and this will make it easier to use spies in
// our test cases
// third, we need to write the test cases and there will be 3 test cases and first up,
// " should render EditExpensePage " and that is an easy one and we will just use a
// snapshot to test that and the next one is " should handle editExpense " and this one
// will require us to use spies and we will be triggering things and making sure the
// spies actually got called and the last one is " should handle removeExpense " and
// and we will be using spies in this test case as well and the first test is similar to the
// first test case in AddExpensePage.test.js and the second and third test are similar to the
// TEST CASE #2 in AddExpensePage.test.js and we can also use beforeEach() if we want to and

// THIS IS OUR NEW CLASS BASED COMPONENT
export class EditExpensePage extends React.Component {

    onSubmit = ( expense ) => {
        // even when using ES6 class properties, we are still able to access component
        // properties if needed ( from Medium article: " Use class properties to clean
        // up your classes and React components " )
        // change this.props.dispatch( editExpense( this.props.expense.id, expense ) );
        // to this.props.editExpense( this.props.expense.id, expense );
        this.props.editExpense( this.props.expense.id, expense );
        // remember, the components we render inside react router get access to a bunch
        // of special props; such as dispatch, history, location and match
        this.props.history.push( '/' );
    };

    onRemove = () => {
        // change this.props.dispatch( removeExpense( id, this.props.expense.id ) );
        // to this.props.removeExpense( id, this.props.expense.id );
        this.props.removeExpense( { id : this.props.expense.id } );
        this.props.history.push( '/' );
    };

    render() {
        return (
            <div>
                <ExpenseForm 
                    expense ={ this.props.expense }
                    onSubmit={ this.onSubmit }
                />
                <button 
                    onClick={ this.onRemove } 
                >Remove
                </button>
            </div>
        );
    }

}
// -- GO TO -- Mark 8 -- BELOW



// THIS IS THE OLD STATELESS FUNCTIONAL COMPONENT
// in lecture 125, we switched to a classed based component and commented out this stateless
// functional component
/*
const EditExpensePage = ( props ) => {
    // console.log( props );
    // -- Mark 5 --
    // delete " Editing expense with the id of { props.match.params.id } " and include
    // the ExpenseForm component and we need to pass in the onSubmit prop to the component
    // and we have access to the expense object just like we had access to it in the
    // AddExpensePage component
    // the next step is to populate the form fields with existing values
    // to get this done we are going to pass in " expense " or type " expense : props.expense "
    // and remember " props.expense " equals the expense object with the correct id, which we
    // derived from the MapStateToProps() function below
    // comment out the console.log from above and were going to set up the expense form to do
    // something meaningful with the expense prop
    // -----------

    // ==============================
    // GO TO EXPENSEFORM.JS -- Mark 6 --
    // ==============================

    return (
        <div>
            <ExpenseForm 
                // -- Mark 6 --
                // so the last thing we need to do is to dispatch the action to edit the page
                // and redirect to the Dashboard page
                // remember, we have to import the editExpense action generator
                // the updates are sitting in the expense argument
                // ----------

                // -- Mark 7 -- from ExpenseListItem
                // challenge
                // add the remove expense button that was copied over from
                // EditExpenseList and and dispatch the removeExpense action generator and
                // redirect to the Dashbaord page
                // remember, we do not have an id variable or { id } but we can set it equal
                // to " props.expense.id " and we have access to that expression
                // remember, we do not have to pass in " props " as the argument to the onSubmit
                // arrow function because already passed in " props " above
                // ------------
                expense ={ props.expense }
                onSubmit={ ( expense ) => {
                    props.dispatch( editExpense( props.expense.id, expense ) );
                    props.history.push( '/' );
                    // console.log( 'updated', expense );
                } }

                // ==============================
                // GO TO EXPENSELISTITEM.JS -- Mark 4 --
                // ==============================
                
            />
            <button 
                onClick={ () => {
                    props.dispatch( removeExpense( { id : props.expense.id } ) );
                    props.history.push( '/' );
                } } 
            >Remove
            </button>
        </div>
    );
};
*/

// way #2 - longhand version
/*
const ExpenseDashBoardPage = () => {
    return (
        <div>
            some text
        </div>
    );
};
*/



// -- Mark 4 --
// we are going to get access to the entire state via the first argument
const mapStateToProps = ( state, props ) => {
    // now, we return an object
    // what do we want off the store?
    // first, we want to be able to give the EditExpensePage component the
    // current expense object
    // now, we're going to be able to access state, which is great because that is
    // where the expenses array lives
    // next, we are going to search the expenses array for an expense whose id matches
    // the " props.match.params.id " above
    // but how do we have access to the props argument?
    // we actually have access to the props argument as the second argument above, which
    // would come after the state argument
    // with the filter method and if the returned value equals true then the item / expense
    // stays in the array and if returned value equals false then the item / expense is removed
    // from the array
    // so we can take some of the current props that were passed into the higher order component
    // and we can use them to calculate new props that we want to include in mapStatetoProps and
    // the EditExpensePage component
    // so react router renders our higher order compoent, which we get from typing
    // " export default connect( mapStateToProps )( EditExpensePage ); "
    // and the HOC passes the props through ( both here in mapStateToProps and in the
    // EditExpensePage component ) and it also allows us to add on some new props ( i.e. expense
    // below )
    // we return an object below
    // were adding one new prop called " expense "
    // so above when we run console.log( props ) within the EditExpensePage component we will see
    // that we have a new prop ( within the props object ) called " expense " and this prop contains
    // the current expense object with all of its key value pairs ( for example, description : "Rent" )
    // in the console we see that the props object contains its own set of props besides the " expense "
    // prop and these other props include dispatch, history, location and match and we are going to
    // be using the props argument passed in as the second argument above to search the expenses array
    return {
        // from MDN: " The find() method returns the value of the first element in the array ( in our case,
        // the first expense item ) that satisfies ( i.e. equals true ) the provided testing function.
        // Otherwise undefined is returned. "
        // the code below will find the expense object ( i.e expense ) that matches the testing function
        // below ( " props.match.params.id === expense.id " ) and we can use " expense " in the
        // EditExpensePage component above
        expense : state.expenses.find( ( expense ) => props.match.params.id === expense.id )
        // also, using filter below resulted in an array of one item being returned ( the right item by the
        // way ) whereas using find resulted in an object being returned
        // why is this?  from stackoverflow: " because array.filter always returns an array (array of all
        // the filtered items), use array.find to get the single object. "
        // expense : state.expenses.filter( ( expense ) => props.match.params.id === expense.id )
    };

};




// -- Mark 8 --
// lecture 125: Testing EditExpensePage
// add ownProps as the second argument below and refer to
// https://react-redux.js.org/using-react-redux/connect-mapdispatch for more information
const mapDispatchToProps = ( dispatch, ownProps ) => ( {
    // editExpense will dispatch whatever comes back from the editExpense action generator and does
    // data need to flow between the argument and the editExpense action generator? yes, we can see
    // from above that we need an id and an expense value
    editExpense   : ( id, expense ) => dispatch( editExpense( id, expense ) ),
    // for remove expense, we will take the ownProps that come in and then pass the ownProps down to
    // the removeExpense() call
    removeExpense : ( ownProps ) => dispatch( removeExpense( ownProps ) )
} );



// now that we connected the EditExpensePage component to the store we have
// access to props.dispatch
// need to add mapDispatchToProps as the second argument below
export default connect( mapStateToProps, mapDispatchToProps )( EditExpensePage );

// END OF -- Mark 8 --


// ==============================
// GO TO -- Mark 5 --
// ==============================