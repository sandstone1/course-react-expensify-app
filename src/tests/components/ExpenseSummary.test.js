

import React from 'react';
// import the named export shallow from enzyme
import { shallow } from 'enzyme';
// remember, to import the named export or unconnected version of the ExpenseListFilters component
import { ExpenseSummary } from '../../components/ExpenseSummary';
// were also going to need the fixture data
import expenses from '../fixtures/expenses';



// TEST CASE #1
test( 'should corrently render ExpenseSummary with 1 expense', () => {
    // remember, since were working with the ExpenseSummary component, we need to pass
    // in expenseCount and expenseTotal
    const wrapper = shallow( <ExpenseSummary expenseCount = { 1 } expenseTotal = { 235 } /> );

    // now that we have our component rendered we will expect something about the wrapper that
    // we just initialized above and we will expect the wrapper to match the snapshot
    expect( wrapper ).toMatchSnapshot();
} );

// if we go to the terminal, we see that this test case is passing


// TEST CASE #2
test( 'should corrently render ExpenseSummary with mulitple expense', () => {
    // remember, since were working with the ExpenseSummary component, we need to pass
    // in expenseCount and expenseTotal
    const wrapper = shallow( <ExpenseSummary expenseCount = { 2 } expenseTotal = { 235 } /> );

    // now that we have our component rendered we will expect something about the wrapper that
    // we just initialized above and we will expect the wrapper to match the snapshot
    expect( wrapper ).toMatchSnapshot();
} );

// if we go to the terminal, we see that this test case is passing


// our snapshot file looks good as well
/*

// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should corrently render ExpenseSummary with 1 expense 1`] = `
<div>
  <h1>
    Viewing 
    1
     
    expense
     totaling 
    $2.35
  </h1>
</div>
`;

exports[`should corrently render ExpenseSummary with mulitple expense 1`] = `
<div>
  <h1>
    Viewing 
    2
     
    expenses
     totaling 
    $2.35
  </h1>
</div>
`;

*/