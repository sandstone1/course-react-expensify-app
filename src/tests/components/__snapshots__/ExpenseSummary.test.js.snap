// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should corrently render ExpenseSummary with 1 expense 1`] = `
<div>
  <h1>
    Viewing 
    1
     
    expense
     totaling 
    $2.35
  </h1>
</div>
`;

exports[`should corrently render ExpenseSummary with mulitple expense 1`] = `
<div>
  <h1>
    Viewing 
    2
     
    expenses
     totaling 
    $2.35
  </h1>
</div>
`;
